src/components/TextInput.js

```diff
@@ -0,0 +1,73 @@
+/*
+This implementation was modeled after the input in
+https://conversation-simple.mybluemix.net/
+ */
+
+import React from 'react';
+import classNames from 'classnames';
+
+const padding = 2;
+
+export default React.createClass({
+  // all html input[type="text"] properties are valid
+  propTypes: {
+    onInput: React.PropTypes.func,
+  },
+
+  getDefaultProps() {
+    return {
+      id: `textinput-${Math.round(Math.random() * 1000)}`,
+      placeholder: 'Type Something',
+    };
+  },
+
+  getInitialState() {
+    return {
+      inputWidth: padding,
+    };
+  },
+
+  onInput(e) {
+    const dummy = e.target.nextSibling;
+    // this is a hack for when you type space,
+    // input text gets shifted because of mismatched width between
+    // dummy and real input.  Char 'r' is used because it has the same
+    // length as a char ' ' in order for the width of the space to count.
+    const text = e.target.value.replace(/\s/g, 'r');
```

Comment:
\s will also match a tab, so the calc isn't right for tabbed text.  `"\t ".match(/\s/)`